cpg_pipes/types.py

```diff
@@ -0,0 +1,224 @@
+        return b.read_input_group(
+            **{
+                'g.vcf.gz': str(self.path),
+                'g.vcf.gz.tbi': str(self.tbi_path),
+            }
+        )
+
+
+FastqPath = Union[str, Path, ResourceFile]
+
+
+@dataclass
+class FastqPair:
+    """
+    Pair of FASTQ files
+    """
+
+    r1: FastqPath
+    r2: FastqPath
+
+    def __getitem__(self, i):
+        assert i == 0 or i == 1, i
+        return [self.r1, self.r2][i]
+
+    def as_resources(self, b) -> 'FastqPair':
+        """
+        Makes a pair of ResourceFile objects for r1 and r2.
+        """
+        r1 = (
+            self.r1 if isinstance(self.r1, ResourceFile) else b.read_input(str(self.r1))
+        )
+        r2 = (
+            self.r2 if isinstance(self.r2, ResourceFile) else b.read_input(str(self.r2))
+        )
+        return FastqPair(r1, r2)
+
+    def __str__(self):
+        return f'{self.r1}|{self.r2}'
+
+
+FastqPairs = List[FastqPair]
+
+
+class SequencingType(Enum):
+    """
+    Type (scope) of a sequencing experiment.
+    """
+
+    GENOME = 'genome'
+    EXOME = 'exome'
+    SINGLE_CELL = 'single_cell'
+    MTSEQ = 'mtseq'
+
+    @staticmethod
+    def parse(str_val: str) -> 'SequencingType':
+        """
+        Parse a string into a SequencingType object.
+        """
+        str_to_val: dict[str, SequencingType] = {} 
+        for val, str_vals in {
+            SequencingType.GENOME: ['genome', 'wgs'],
+            SequencingType.EXOME: ['exome', 'wts', 'wes'],
+            SequencingType.SINGLE_CELL: ['single_cell', 'single_cell_rna'],
+            SequencingType.MTSEQ: ['mtseq']
+        }.items():
+            for str_v in str_vals:
+                str_v = str_v.lower()
+                assert str_v not in str_to_val, (str_v, str_to_val)
+                str_to_val[str_v] = val
+        str_val = str_val.lower()
+        if str_val not in str_to_val:
+            raise ValueError(
+                f'Unrecognised sequence type {str_val}. '
+                f'Available: {list(str_to_val.keys())}'
+            )
+        return str_to_val[str_val]
+
+
+class AlignmentInput:
+    """
+    Alignment input can be an indexed CRAM file, or a list of FASTQ file pairs.
+    """
+    def __init__(
+        self, 
+        data: Union[CramPath, FastqPairs],
+        sequencing_type: SequencingType,
+    ):
+        self.data = data
+        self.is_cram_or_bam = isinstance(data, CramPath)
+        self.sequencing_type = sequencing_type
+        
+    def __str__(self):
+        return f'{self.sequencing_type.value}:{str(self.data)}'
+
+    def exists(self) -> bool:
+        """
+        Check if all files in alignment inputs exist.
+        """
+        if isinstance(self.data, CramPath):
+            return utils.exists(self.data.path)
+        else:
+            return all(
+                utils.exists(pair.r1) and utils.exists(pair.r2) 
+                for pair in self.data
+            )
+
+    def compact_file_paths(self) -> str:
```

Comment:
Could we refactor this out to it's own method, looks useful :D
A set of doctests for the future would be useful here too.

Reply:
I refactored it a bit in the other PR after your suggestions. Going to update it here now as well :)